src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonNode.cs

```diff
@@ -232,6 +239,112 @@ public JsonNode Root
+        public JsonNode DeepClone()
+        {
+            return InternalDeepClone();
+        }
+
+        internal abstract JsonNode InternalDeepClone();
+
+        /// <summary>
+        /// Returns <see cref="JsonValueKind"/> of current instance.
+        /// </summary>
+        public JsonValueKind GetValueKind()
+        {
+            return this switch
+            {
+                JsonObject => JsonValueKind.Object,
+                JsonArray => JsonValueKind.Array,
+                _ => AsValue().GetInternalValueKind(),
+            };
+        }
+
+        /// <summary>
+        /// Returns property name of the current node from the parent object.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">
+        /// The current parent is not a <see cref="JsonObject"/>.
+        /// </exception>
+        public string GetPropertyName()
+        {
+            JsonObject? jsonObject = _parent as JsonObject;
+
+            if (jsonObject is null)
+            {
+                ThrowHelper.ThrowInvalidOperationException_NodeWrongType(nameof(JsonObject));
+            }
+
+            return jsonObject.GetPropertyName(this);
+        }
+
+        /// <summary>
+        /// Returns index of the current node from the parent <see cref="JsonArray" />.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">
+        /// The current parent is not a <see cref="JsonArray"/>.
+        /// </exception>
+        public int GetElementIndex()
+        {
+            JsonArray? jsonArray = _parent as JsonArray;
+
+            if (jsonArray is null)
+            {
+                ThrowHelper.ThrowInvalidOperationException_NodeWrongType(nameof(JsonArray));
+            }
+
+            return jsonArray.GetElementIndex(this);
+        }
+
+        /// <summary>
+        /// Compares the values of two nodes, including the values of all descendant nodes.
+        /// </summary>
+        /// <param name="node1">The <see cref="JsonNode"/> to compare.</param>
+        /// <param name="node2">The <see cref="JsonNode"/> to compare.</param>
+        /// <returns><c>true</c> if the tokens are equal; otherwise <c>false</c>.</returns>
+        public static bool DeepEquals(JsonNode? node1, JsonNode? node2)
+        {
+            if (node1 is null)
+            {
+                return node2 is null;
+            }
+
+            return node1.DeepEquals(node2);
+        }
+
+        internal abstract bool DeepEquals(JsonNode? node);
+
+        /// <summary>
+        /// Replaces this node with a new value.
+        /// </summary>
+        /// <typeparam name="T">The type of value to be replaced.</typeparam>
+        /// <param name="value">Value that replaces this node.</param>
+        [RequiresUnreferencedCode(JsonValue.CreateUnreferencedCodeMessage)]
+        [RequiresDynamicCode(JsonValue.CreateDynamicCodeMessage)]
+        public void ReplaceWith<T>(T value)
+        {
+            if (_parent is null)
```

Comment:
The `_parent` field is being read multiple times in this method. Perhaps refactor into a switch statement/expression?